drivers/can/can_mcp251xfd.c

```diff
@@ -0,0 +1,1814 @@
+	const struct device *dev = dev_data->dev;
+	const struct mcp251xfd_config *dev_cfg = dev->config;
+	int ret;
+
+	/* Disable pin interrupts */
+	ret = gpio_pin_interrupt_configure_dt(&dev_cfg->int_gpio_dt, GPIO_INT_DISABLE);
+	if (ret < 0) {
+		LOG_ERR("Couldn't disable pin interrupt [%d]", ret);
+		k_oops();
+	}
+
+	k_sem_give(&dev_data->int_sem);
+}
+
+static int mcp251xfd_get_capabilities(const struct device *dev, can_mode_t *cap)
+{
+	ARG_UNUSED(dev);
+
+	*cap = CAN_MODE_NORMAL | CAN_MODE_LISTENONLY | CAN_MODE_LOOPBACK;
+
+#if defined(CONFIG_CAN_FD_MODE)
+	*cap |= CAN_MODE_FD;
+#endif
+
+	return 0;
+}
+
+static int mcp251xfd_start(const struct device *dev)
+{
+	struct mcp251xfd_data *dev_data = dev->data;
+	const struct mcp251xfd_config *dev_cfg = dev->config;
+	int ret;
+
+	if (dev_data->started) {
+		return -EALREADY;
+	}
+
+	/* in case of a race between mcp251xfd_send() and mcp251xfd_stop() */
+	mcp251xfd_reset_tx_fifos(dev, -ENETDOWN);
+
+	if (dev_cfg->phy != NULL) {
+		ret = can_transceiver_enable(dev_cfg->phy);
+		if (ret < 0) {
+			LOG_ERR("Failed to enable CAN transceiver [%d]", ret);
+			return ret;
+		}
+	}
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+
+	ret = mcp251xfd_set_mode_internal(dev, dev_data->next_mcp251xfd_mode);
+	if (ret < 0) {
+		LOG_ERR("Failed to set the mode [%d]", ret);
+		if (dev_cfg->phy != NULL) {
+			/* Attempt to disable the CAN transceiver in case of error */
+			(void)can_transceiver_disable(dev_cfg->phy);
+		}
+	} else {
+		dev_data->started = true;
+	}
+
+	k_mutex_unlock(&dev_data->mutex);
+
+	return ret;
+}
+
+static int mcp251xfd_stop(const struct device *dev)
+{
+	struct mcp251xfd_data *dev_data = dev->data;
+	const struct mcp251xfd_config *dev_cfg = dev->config;
+	uint8_t *reg_byte;
+	int ret;
+
+	if (!dev_data->started) {
+		return -EALREADY;
+	}
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+
+	/* abort all transmissions */
+	reg_byte = mcp251xfd_get_spi_buf_ptr(dev);
+	*reg_byte = MCP251XFD_UINT32_FLAG_TO_BYTE_MASK(MCP251XFD_REG_CON_ABAT);
+
+	ret = mcp251xfd_write(dev, MCP251XFD_REG_CON_B3, 1);
+	if (ret < 0) {
+		k_mutex_unlock(&dev_data->mutex);
+		return ret;
+	}
+
+	/* wait for all the messages to be aborted */
+	while (1) {
+		reg_byte = mcp251xfd_read_crc(dev, MCP251XFD_REG_CON_B3, 1);
+
+		if (!reg_byte ||
+		    (*reg_byte & MCP251XFD_UINT32_FLAG_TO_BYTE_MASK(MCP251XFD_REG_CON_ABAT)) == 0) {
+			break;
+		}
+	}
+
+	mcp251xfd_reset_tx_fifos(dev, -ENETDOWN);
+
+	ret = mcp251xfd_set_mode_internal(dev, MCP251XFD_REG_CON_MODE_CONFIG);
+	if (ret < 0) {
+		k_mutex_unlock(&dev_data->mutex);
+		return ret;
+	}
+
+	dev_data->started = false;
+	k_mutex_unlock(&dev_data->mutex);
+
+	if (dev_cfg->phy != NULL) {
+		ret = can_transceiver_disable(dev_cfg->phy);
+		if (ret < 0) {
+			LOG_ERR("Failed to disable CAN transceiver [%d]", ret);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static const struct can_driver_api mcp251xfd_api_funcs = {
```

Comment:
Please move these further down in the file similar to other driver implementations.

Reply:
thanks, updated.